### ISSUE TITLE
[expo-auth-session] Popup doesn't close after redirect using React Hooks

### ISSUE THREAD

**Issue by simoneb** [expo/expo#8134]
### Environment

Expo Client and Web

### Steps to Reproduce

- Use expo-auth-session Hooks as shown in the guides at this page: https://docs.expo.io/versions/latest/sdk/auth-session
- Invoke the `promptAsync` function returned by `useAuthRequest`
- A popup browser window opens which allows to authenticate to the external provider
- In Web: the application is reloaded inside the popup rather than closing the popup. Authentication information is provided to the app as querystring parameters to the `redirectUrl`, yet the app doesn't seem to be picking them up
- In Client: an error appears which says:

> Something went wrong trying to finish signing in. Please close this screen to go back to the app

### Expected Behavior

The popup closes automatically and the user is authenticated

### Actual Behavior

In both Web and Client, closing the popup/screen will cause the auth result to be "dismissed"

### Reproducible Demo

I don't have one available but I can create one if useful. I'd like to understand if there's anything I'm doing clearly wrong and this is supposed to work first. None of the examples in the examples repository show Hooks being used.

**Comment by byCedric** [comment]
Hi @simoneb! It would definitely be helpful to add a small repro, e.g. in snack. 😄 I'm not sure what's going on, but maybe you can check out some of the vendor-specific examples. It should include an example with the hooks you mentioned, e.g. [GitHub](https://docs.expo.io/versions/latest/sdk/auth-session/#github) or [Google](https://docs.expo.io/versions/latest/sdk/auth-session/#google).

**Comment by simoneb** [comment]
Found the solution, which is documented only in the [Identity 4](https://docs.expo.io/guides/authentication/#identity-4) section of the docs, which is basically adding this code:

```
if (Platform.OS === 'web') {
  WebBrowser.maybeCompleteAuthSession();
}
```

**Comment by sbrighiu** [comment]
I don't follow @simoneb where do you add that code? And how will it force the AuthSession auth methods to return the resulting url?

**Comment by simoneb** [comment]
> I don't follow @simoneb where do you add that code? And how will it force the AuthSession auth methods to return the resulting url?

Anywhere in the app as long as it executes when the login process redirects back to the app. Executing that code will close the login popup.

**Comment by sbrighiu** [comment]
My problem is that the popup does not close and instead opens the redirect url inside it. Closing the window doesnt help me since the main window dod not receive the payload

**Comment by simoneb** [comment]
> My problem is that the popup does not close and instead opens the redirect url inside it. Closing the window doesnt help me since the main window did not receive the payload

Then you may be having a different issue

**Comment by sbrighiu** [comment]
Could be :) Thanks for clarifying :)

**Comment by pakaplace** [comment]
@sbrighiu were you able to fix your issue the the redirect URL opening inside the popup? I have the same issue

**Comment by sbrighiu** [comment]
Yup. There was a global callback that needed to be implemented in index.js so that when the popup would try to open the page it would detect it and close. Read the documentation. It’s a one liner declared globally before app component.

--------------------------------------
Stefan Brighiu
Software Engineer @ AB4 Systems
Phone: +40 744 581 921
Web: http://ab4.systems

**Comment by sbrighiu** [comment]
import * as WebBrowser from 'expo-web-browser';
if (Platform.OS === 'web') {
  WebBrowser.maybeCompleteAuthSession();
}
@pakaplace

**Comment by bzozoo** [comment]
I have same issue

**Comment by aadityataparia** [comment]
Same issue, popup doesn't close but it happens only in chrome extension (I am opening my website as iframe in chrome extension). Popup redirects back to home page, but doesn't close and chrome extension doesn't receive payload either. I have already added `WebBrowser.maybeCompleteAuthSession()`

**Comment by MehmetKaplan** [comment]
Aditya, could you have solved this?

**Comment by MehmetKaplan** [comment]
Thanks to @sbrighiu 's comment, my case is solved.

In my case the strange thing was (and partially still is), everything was working fine in the localhost but when I deployed the app to a server (after building web with the command `npx expo export -p web`) it was failing.

And coming back to the `WebBrowser.maybeCompleteAuthSession();` statement, it already existed in a file named `GoogleLogin.js`. 

When I placed the `WebBrowser.maybeCompleteAuthSession();` to the `app/index.js` file, everything started work like a charm:

```javascript
// app/index.js
import 'expo-dev-client';
import { Redirect } from 'expo-router';
import * as WebBrowser from "expo-web-browser";
WebBrowser.maybeCompleteAuthSession();

export default function Page() {
	return <Redirect href={"/app"} />;
}
```

So my conclusion is, somehow the `WebBrowser.maybeCompleteAuthSession();` is being run when I do `yarn web` but it is not done when I do `npx expo export -p web`. And the difference may be the `-p` parameter. 

My guess, in production mode the `GoogleLogin.js` is not touched during the redirect url rendering in the popup, but in test mode it is touched. (_**Speculation**_: Can be related with react revisiting objects twice in the test mode.)
